File: 1379. Find a Corresponding Node of a Binary Tree in a Clone of That Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
  TreeNode* dfs(TreeNode* original, TreeNode* cloned, TreeNode* target){
    if (original == NULL){
      return NULL;
    }
    if (original == target){
      return cloned;
    }
    TreeNode* left = dfs(original->left, cloned->left, target);
    TreeNode* right = dfs(original->right, cloned->right, target);
    if (left){
      return left;
    }
    else if (right){
      return right;
    }
    return NULL;
  }
public:
  TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
    return dfs(original, cloned, target);
  }
};

File: 1379. Find a Corresponding Node of a Binary Tree in a Clone of That Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
  TreeNode dfs(TreeNode original, TreeNode cloned, TreeNode target){
    if (original == null){
      return null;
    }
    if (original == target){
      return cloned;
    }
    TreeNode left = dfs(original.left, cloned.left, target);
    TreeNode right = dfs(original.right, cloned.right, target);
    if (left != null){
      return left;
    }
    else if (right != null){
      return right;
    }
    return null;
  }
  public final TreeNode getTargetCopy(final TreeNode original, final TreeNode cloned, final TreeNode target) {
    return dfs(original, cloned, target);
  }
}